### ISSUE TITLE
Docker packaging tests for ARM

### ISSUE THREAD

**Issue by pugnascotia** [elastic/elasticsearch#54925]
We have a suite of tests for our Docker image in the `DockerTests` class, which can be run via:

    ./gradlew :qa:os:destructiveDistroTest.default-docker

This task is configured to depend on a Docker export task, which itself depends on a Docker build task. This works fine. However following #53936, now we can build Docker images for ARM, with:

    ./gradlew distribution:docker:buildAarch64DockerImage

but there is no support for running the packaging tests against the ARM image, because the dependency graph assumes the x86 build.

Now, the Docker packaging tests don't actually care what Docker machine's architecture is. We could make the tests depend on an "alias" task, which builds whichever is the correct Docker image for the current platform.

In so doing, we would also give ourselves a point to override this behaviour, and explicitly target a specific architecture. The use-case is where someone is developing the Docker images and has a Docker machine capable of running more x86 and ARM images (notably, Docker for Mac et al support this via QEMU and `binfmt_misc`). In this case, it would be useful to be able to perform e.g.

    ./gradlew -Ddocker.arch=aarch64 :qa:os:destructiveDistroTest.default-docker

Orthogonal but related to this is that for [multi-arch Docker builds](https://www.docker.com/blog/multi-arch-images/), we will need to be able to build and push images with different tags, which are then brought together using `docker manifest`. Right now, the ARM Docker build creates images with the same tags as the x86 images, which is pretty confusing when you're working on the images locally. We should consider including the architecture in the Docker tag, for both `x86` and `aarch64`. In this case it might be useful to use Docker's conventions, which are `amd64` and `arm64`.

FYI @mark-vieira @jasontedor

**Comment by mark-vieira** [comment]
Looking at this again we already sort of have such alias tasks. The behavior though is that it runs _all_ tests, ignoring those not applicable to the current platform. So just `destructiveDistroTest` will run all packaging tests, skipping those that don't apply to the current machine (ex: skips docker tests on Windows).

We have more specific versions of these tasks as well, such as `destructiveDistroTest.docker`, that runs _only _ the Docker tests. Right now this does both default and oss distributions but we could just wire up the arm ones as well, and on a non-arm system these would be skipped (unless manually overriden, in which case all tests would run).

@pugnascotia Would this situation work? Would just continuing the existing pattern of "skip stuff that doesn't apply" align with this or do we need something different?

In what I'm running `./gradlew destructiveDistroTest.docker` on an x86_64 system would skip the arm tests and vice versa on an arm system. This would still rely on tasks per-architecture so we'd now have `destructiveDistroTest.default-docker-x86_64` and `destructiveDistroTest.default-docker-aarch64`. I don't think we can avoid this because we'll want to do the same for the other packaging types as well (deb, rpm, tar), not just Docker.

**Comment by rjernst** [comment]
>  because we'll want to do the same for the other packaging types as well (deb, rpm, tar), not just Docker.

In fact, we already have these tasks for every distribution, right now, it is just that like Mark said, we assume all are x86. We should just need another loop in DistroTestPlugin.

```
Verification tasks
------------------
distroTest.default-deb - Runs default-deb-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.default-deb-no-jdk - Runs default-deb-no-jdk-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.default-docker - Runs default-docker-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.default-linux-archive - Runs default-linux-archive-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.default-linux-archive-no-jdk - Runs default-linux-archive-no-jdk-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.default-rpm - Runs default-rpm-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.default-rpm-no-jdk - Runs default-rpm-no-jdk-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.oss-deb - Runs oss-deb-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.oss-deb-no-jdk - Runs oss-deb-no-jdk-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.oss-docker - Runs oss-docker-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.oss-linux-archive - Runs oss-linux-archive-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.oss-linux-archive-no-jdk - Runs oss-linux-archive-no-jdk-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.oss-rpm - Runs oss-rpm-8.0.0-SNAPSHOT distribution tests within vagrant
distroTest.oss-rpm-no-jdk - Runs oss-rpm-no-jdk-8.0.0-SNAPSHOT distribution tests within vagrant
```

**Comment by pugnascotia** [comment]
OK, so if I add in an extra loop for architecture, and also a way of say "No, I really want $ARCH for the Docker stuff", then that would probably get us most of the way there. Thanks all.

**Comment by james-crowley** [comment]
> We should consider including the architecture in the Docker tag, for both x86 and aarch64. In this case it might be useful to use Docker's conventions, which are amd64 and arm64.

Naming/tagging the docker containers with those docker’s conventions make the most sense!

In the code to add support for `aarch64`, I keep seeing either switch statements or ternary operators. Especially, when you have the situations like you described above where you need to convert the name `aarch64` into docker’s way of naming things. This also happens when trying to produce `deb` and `rpm` packages since both have unique naming schemes for architectures.

One idea I wanted to run by everyone is moving some of these switch statements and ternary operators into the [`Architecture.java`](https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/java/org/elasticsearch/gradle/Architecture.java) file. Seems like there is already one method inside the file, `current()`. 
https://github.com/elastic/elasticsearch/blob/cbc9c7ab4dd6833dc9985c25ea23f4990c2b539d/buildSrc/src/main/java/org/elasticsearch/gradle/Architecture.java#L27-L38

We could add additional methods to support our needs.

Bringing most of the architecture logic into [`Architecture.java`](https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/java/org/elasticsearch/gradle/Architecture.java) would help reduce duplicate code. Implementing a method that gives back docker’s naming convention of the architecture would be helpful in many different places throughout the code. 

@jasontedor @rjernst @mark-vieira Thoughts? Happy to post back here with a code snippet of what I am thinking or do a PR if you guys want.

**Comment by jasontedor** [comment]
The point of that class is only to serve as a place to obtain the current architecture, which is needed in various places in the build. I worry about overloading to also serve as a place to house logic for packaging, which is different for Docker/RPM/Deb, as you point out.

**Comment by pugnascotia** [comment]
This issue was actually addressed in #55073, as the build now constructs testing targets such as `:qa:os:destructiveDistroTest.default-docker-aarch64`. I'm no longer convinced of the need to cross-test the images, so I'm going to close this issue for the time being.
